README.md
# Balances

A Clojure library designed to simulate bank operations.

## Solution

The main variable is commonly called **ops**. It is a map where the keys are the account number and the value is a vector of Operations for that account.
Operation is a record with three attributes: description, amount and date.

1. description: A short string describing the operation.
2. amount: if the operation was credit, the amount must be positive; if it was debit, the amount must be negative; it's supposed that the sign is coherent with the description.
3. date: the date when the operation occurred, in the format dd/MM.

The solution has three layers, each one in its namespace:

* util: simple, common and pure functions; stateless namespace.
* core: main functions to operate over **ops**; the functions are also pure and the namespace is stateless.
* server: HTTP endpoints; this is the only layer with state which has the **ops** variable to hold all past operations of all accounts.

## Main functions

### new-operation
Creates a new Operation instance and update the map **ops**.
To update **ops**, the instance is conjoined with the vector that is the value for the account number key.

### current-balance
All operations of an account can be retrieved from **ops** by passing the account number as key.
Having these operations, the current balance is calculated by summing all signed amounts from all operations since the beginning.
 
### bank-statement
This function gets the log of operations of an account between two dates, namely **start date** and **end date**.
First, the balance is calculated for all dates at that account.
Then, the Operations are filtered to get those between **start date** and **end date**.
Finally, at each of these dates a new key-value is associated in the map: 

* key: date as string.
* value: map with the operations' descriptions and the balance.

### debt-periods
A vector is constructed. Each element of the vector is a map. The map contains a **start date**, a **principal** and maybe an **end date**.

First, the balance is calculated for all dates at the specified account.

This sequence is sorted by date.

Each new date that has a negative balance is added to the vector above.

If the balance becomes positive in a following date, the last element of the vector is updated with an **end date**.

If the balance changes but continues negative, then the last element is updated with an **end date** and a new element is added with a **start date** and a **principal**.

## Hypothesis

* The user that adds new operations is authenticated and using a secure connection (HTTPS).
* Following the exercise, it is not considered the year of the operations, but it's a simple extension that could be done in ns **util**.
* The user already has a valid account number over which he/she operates.
* The amount's sign and the description are coherent; for example, if the operation is a credit, the amount is positive, if it is a purchase, the amount is negative. This could be guaranteed by using a new parameter or checking from the description.

## Decisions

* The account number can be of any type: integer, UUID, string, etc. The tests use integer for the sake of readability.
* All HTTP requests are POST so they can accept JSON payloads as requests. 
* It is used double for the amounts instead of float for better truncations.
* It is used record instead of maps for better performance to access values and to standardize its attributes.

## Usage

To run as localhost at port 3000, execute:

```shell
lein run
```

To test all namespaces, execute:
```shell
lein test
```

To test a specific namespace add its name after, for example:
```shell
lein test balances.core-test
```

## License

Copyright © 2016 Henrique Rodrigues

Distributed under the MIT License.

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2016-11-16
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2016-11-16
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/balances/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/balances/compare/0.1.0...0.1.1


test/balances/core_test.clj
(ns balances.core-test
  (:require [clojure.test :refer [deftest testing is]]
            [balances.core :refer [new-operation current-balance build
                                   bank-statement debt-periods]])
  (:import (balances.core Operation)))

(defn- opp
  "Operation params: build a map more succinctly"
  [account description amount date]
  {:account account :description description :amount amount :date date})

;;;; HELPER FUNCTIONS TESTS
(deftest all-dates->string-test
  (let [operations [(build "Debit" -146.32 "06/12")
                    (build "Credit" 370.05 "12/12")
                    (build "Purchase" -130.44 "01/12")]
        str-dates (#'balances.core/all-date->string operations)
        result [(Operation. "Debit" -146.32 "06/12")
                (Operation. "Credit" 370.05 "12/12")
                (Operation. "Purchase" -130.44 "01/12")]]
    (is (= result str-dates))))

(deftest compute-balances-test
  (let [ops {1 [(build "Credit" 1000.00 "07/12")
                (build "Debit" -146.32 "06/12")
                (build "Purchase" -32.20 "06/12")]
             2 [(build "Deposit" 500.00 "06/12")
                (build "Credit" 1000.00 "07/12")
                (build "Salary" 4267.34 "06/12")]}
        balances-1 (#'balances.core/compute-balances ops 1)
        balances-2 (#'balances.core/compute-balances ops 2)
        result-1 [{:date "06/12" :balance -178.52}
                  {:date "07/12" :balance 821.48}]
        result-2 [{:date "06/12" :balance 4767.34}
                  {:date "07/12" :balance 5767.34}]]
    (is (= balances-1 result-1))
    (is (= balances-2 result-2))))

(deftest compute-each-balance-test
  (let [ops {1 [(build "Credit" 1000.00 "07/12")
                (build "Debit" -146.32 "06/12")
                (build "Purchase" -32.20 "06/12")]
             2 [(build "Deposit" 500.00 "06/12")
                (build "Credit" 1000.00 "07/12")
                (build "Salary" 4267.34 "06/12")]}
        balances-1 (#'balances.core/compute-each-balance ops 1)
        balances-2 (#'balances.core/compute-each-balance ops 2)
        result-1 {"06/12" -178.52
                  "07/12" 821.48}
        result-2 {"06/12" 4767.34
                  "07/12" 5767.34}]
    (is (= balances-1 result-1))
    (is (= balances-2 result-2))))


;;;; NEW OPERATION TESTS
(deftest new-operation-test
  (let [opr-1 (opp 1 "Deposit" 1000.00 "15/10")
        opr-2 (opp 1 "Purchase on Amazon" -800.00 "17/10")
        opr-3 (opp 2 "Debit" -199.43 "17/11")
        opr-4 (opp 3 "Salary" 8123.30 "10/10")]

    (testing "Initial case"
      (let [ops (new-operation {} opr-1)
            res {1 [(build "Deposit" 1000.00 "15/10")]}]
        (is (= ops res))))

    (testing "Add new operation"
      (let [ops (-> (new-operation {} opr-1) (new-operation opr-2))
            res {1 [(build "Purchase on Amazon" -800.00 "17/10")
                    (build "Deposit" 1000.00 "15/10")]}]
        (is (= ops res))))

    (testing "Add multiple accounts"
      (let [ops (reduce new-operation {} [opr-1 opr-2 opr-3 opr-4])
            res {1 [(build "Purchase on Amazon" -800.00 "17/10")
                    (build "Deposit" 1000.00 "15/10")]
                 2 [(build "Debit" -199.43 "17/11")]
                 3 [(build "Salary" 8123.30 "10/10")]}]
        (is (= ops res))))))


;;;; CURRENT BALANCE TESTS
(deftest current-balance-test
  (let [operations [(opp 1 "Deposit" 1423.34 "15/10")
                    (opp 1 "Purchase on Amazon" -898.76 "17/10")
                    (opp 2 "Salary" 8115.40 "12/10")
                    (opp 2 "Debit to Mary" -100.00 "14/10")]
        ops (reduce new-operation {} operations)]

    (testing "Current balance of account number 1"
      (is (= (current-balance ops 1) {:balance 524.57})))

    (testing "Current balance of account number 2"
      (is (= (current-balance ops 2) {:balance 8015.40})))

    (testing "Current balance of absent account"
      (is (nil? (current-balance ops 3))))))


;;;; BANK STATEMENT TESTS
(deftest bank-statement-test
  (let [operations [(opp 1 "Deposit" 1000.00 "15/10")
                    (opp 1 "Purchase on Amazon" -800.00 "17/10")
                    (opp 2 "Salary" 8000.00 "15/10")
                    (opp 2 "Debit to Mary" -100.00 "17/10")]]

    (testing "statement with one operation for account 1"
      (let [sta (bank-statement
                  (new-operation {} (operations 0)) 1 "15/10" "17/10")
            res {"15/10" {:operations ["- Deposit 1000.00"]
                          :balance 1000.00}}]
        (is (= sta res))))

    (testing "statement with two operation for account 1"
      (let [sta (bank-statement
                  (reduce new-operation {} (take 2 operations)) 1 "15/10" "17/10")
            res {"15/10" {:operations ["- Deposit 1000.00"]
                          :balance 1000.00}
                  "17/10" {:operations ["- Purchase on Amazon 800.00"]
                           :balance 200.00}}]
        (is (= sta res))))

    (testing "statement with three operation for account 2"
      (let [sta (bank-statement
                  (reduce new-operation {} (take 3 operations)) 2 "15/10" "17/10")
            res {"15/10" {:operations ["- Salary 8000.00"]
                          :balance 8000.00}}]
        (is (= sta res))))

    (testing "statement with four operation for account 2"
      (let [sta (bank-statement
                  (reduce new-operation {} operations) 2 "15/10" "17/10")
            res {"15/10" {:operations ["- Salary 8000.00"]
                          :balance 8000.00}
                  "17/10" {:operations ["- Debit to Mary 100.00"]
                           :balance 7900.00}}]
        (is (= sta res))))

    (testing "statement for absent account"
      (let [sta (bank-statement
                  (reduce new-operation {} operations) 3 "15/10" "17/10")
            res {}]
        (is (= sta res))))))


(deftest bank-statement-between-dates-test
  (let [operations [(opp 1 "Deposit" 1000.00 "15/10")
                    (opp 1 "Purchase" -1200.00 "16/10")
                    (opp 1 "Salary" 4000.00 "19/10")
                    (opp 1 "Debit" -20000.00 "21/10")]
        res {"15/10" {:balance 1000.00   :operations ["- Deposit 1000.00"]}
             "16/10" {:balance -200.00   :operations ["- Purchase 1200.00"]}
             "19/10" {:balance 3800.00   :operations ["- Salary 4000.00"]}
             "21/10" {:balance -16200.00 :operations ["- Debit 20000.00"]}}
        ops (reduce new-operation {} operations)]

    (testing "Statement within dates"
      (let [sta (bank-statement ops 1 "15/10" "21/10")
            res# res]
        (is (= sta res#))))

    (testing "Statement that overlaps dates"
      (let [sta (bank-statement ops 1 "10/10" "16/10")
            res# (select-keys res ["15/10" "16/10"])]
        (is (= sta res#))))

    (testing "Statement that overlaps dates"
      (let [sta (bank-statement ops 1 "16/10" "29/10")
            res# (dissoc res "15/10")]
        (is (= sta res#))))))


(deftest bank-statement-multiple-operations-test
  (let [operations [(opp 1 "Deposit" 1000.00 "15/10")
                    (opp 1 "Purchase" -1200.00 "16/10")
                    (opp 1 "Salary" 4000.00 "16/10")
                    (opp 1 "Debit" -2000.00 "16/10")
                    (opp 1 "Credit" 5000.00 "19/10")
                    (opp 1 "Withdrawal" -3800.00 "19/10")]
        res {"15/10" {:balance 1000.00 :operations ["- Deposit 1000.00"]}
             "19/10" {:balance 3000.00 :operations ["- Withdrawal 3800.00"
                                                    "- Credit 5000.00"]}
             "16/10" {:balance 1800.00 :operations ["- Debit 2000.00"
                                                    "- Salary 4000.00"
                                                    "- Purchase 1200.00"]}}
        ops (reduce new-operation {} operations)]

    (testing "Multiple operations on 16/10 and 19/10"
      (let [sta (bank-statement ops 1 "15/10" "19/10")]
        (is (= sta res))))))


;;;; DEBT PERIODS TEST
(deftest debt-periods-test
  (let [ops [(opp 1 "Deposit"    1000.00 "15/10")  ; 1000
             (opp 1 "Purchase"  -1200.00 "16/10")
             (opp 1 "Salary"     4000.00 "16/10")  ; 3800
             (opp 1 "Withdrawal"-5000.00 "19/10")  ; -1200
             (opp 1 "Credit"     3000.00 "21/10")]]; 1800

    (testing "One period of debt with end"
      (let [debts (debt-periods (reduce new-operation {} ops) 1)
            res {:debts [{:start "19/10" :end "20/10" :principal -1200.00}]}]
        (is (= debts res))))

    (testing "Two periods of debt with ends"
      (let [ops# (conj ops (opp 1 "Debit" -2000.00 "23/10") ; -200
                           (opp 1 "Credit" 3000.00 "26/10")); 2800
            res {:debts [{:start "23/10" :end "25/10" :principal -200.00}
                         {:start "19/10" :end "20/10" :principal -1200.00}]}
            debts (debt-periods (reduce new-operation {} ops#) 1)]
        (is (= debts res))))

    (testing "Period of debt without end"
      (let [debts (debt-periods (reduce new-operation {} (drop-last ops)) 1)
            res {:debts [{:start "19/10" :principal -1200.00}]}]
        (is (= debts res))))))


(deftest debt-periods-same-debt-test
  (testing "Periods with debt when principal doesn't change"
    (let [ops [(opp 1 "Debit"    -100.00  "12/01")
               (opp 1 "Credit"    200.00  "15/01")
               (opp 1 "Purchase" -200.00  "15/01")
               (opp 1 "Salary"    1100.00 "20/01")]
          res {:debts [{:start "12/01" :principal -100.00 :end "19/01"}]}
          debts (debt-periods (reduce new-operation {} ops) 1)]
      (is (= debts res)))))


(deftest debt-periods-multiple-accounts-test
  (let [operations [(opp 1 "Deposit"     1000.00 "15/10")
                    (opp 1 "Purchase"   -1200.00 "16/10")
                    (opp 2 "Salary"      4000.00 "16/10")
                    (opp 2 "Withdrawal" -5000.00 "19/10")
                    (opp 1 "Credit"      3000.00 "21/10")]
        ops (reduce new-operation {} operations)]

    (testing "Debt periods of account 1"
      (let [debts (debt-periods ops 1)
            res {:debts [{:start "16/10" :end "20/10" :principal -200.00}]}]
        (is (= debts res))))

    (testing "Debt periods of account 2"
      (let [debts (debt-periods ops 2)
            res {:debts [{:start "19/10" :principal -1000.00}]}]
        (is (= debts res))))))


(deftest debt-periods-no-debt
  (let [operations [(opp 1 "Deposit"    1000.00 "15/10")
                    (opp 1 "Purchase"   1200.00 "16/10")
                    (opp 2 "Salary"     4000.00 "16/10")
                    (opp 2 "Withdrawal" 5000.00 "19/10")
                    (opp 1 "Credit"     3000.00 "21/10")]
        ops (reduce new-operation {} operations)]

    (testing "Account 1 with no debt periods"
      (let [debts (debt-periods ops 1)
            res {:debts []}]
        (is (= debts res))))

    (testing "Account 2 with no debt periods"
      (let [debts (debt-periods ops 2)
            res {:debts []}]
        (is (= debts res))))))

test/balances/util_test.clj
(ns balances.util-test
  (:require [balances.util :as util]
            [clojure.test :refer [deftest testing is]]))

(deftest previous-day-test
  (testing "day in the middle of a month"
    (is (= "16/10" (util/previous-day "17/10"))))

  (testing "day at the beginning of a month"
    (is (= "30/11" (util/previous-day "01/12"))))

  (testing "day at the beginning of a year"
    (is (= "31/12" (util/previous-day "01/01")))))

(deftest next-day-test
  (testing "day in the middle of a month"
    (is (= "16/10" (util/date->string (util/next-day "15/10")))))

  (testing "day in the end of a month"
    (is (= "01/02" (util/date->string (util/next-day "31/01")))))

  (testing "day in the end of a year"
    (is (= "01/01" (util/date->string (util/next-day "31/12"))))))

(deftest abs-test
  (testing "for integers"
    (is (= 10 (util/abs 10))
        (= 10 (util/abs -10))))

  (testing "for double"
    (is (= 3.14159 (util/abs 3.14159))
        (= 3.14159 (util/abs -3.14159))))

  (testing "for ratio"
    (is (= 20/7 (util/abs 20/7))
        (= 20/7 (util/abs -20/7))))

  (testing "for BigInt"
    (is (= 100N (util/abs 100N))
        (= 100N (util/abs -100N))))

  (testing "for BigDecimal"
    (is (= 2718.281M (util/abs 2718.281M))
        (= 2718.281M (util/abs -2718.281M))))

  (testing "for string"
    (is (thrown? Exception (util/abs "3")))))

(deftest to-float-test
  (testing "for integer"
    (is (= 1.00 (util/to-format 1))
        (= -1.00 (util/to-format -1))))

  (testing "for float"
    (is (= 100.15 (util/to-format 100.1515151))
        (= -100.15 (util/to-format -100.15131341))))

  (testing "for strings"
    (is (= 100.50 (util/to-format "100.50292929"))
        (= -100.50 (util/to-format "-100.50278723")))))

test/balances/server_test.clj
(ns balances.server-test
  (:require [clojure.test :refer [is deftest testing]]
            [balances.server :refer [ops app]]
            [balances.core :refer [build]]
            [ring.mock.request :as mock]
            [clojure.data.json :as json]))

(defn- opp
  "Operation params: build a map more succinctly"
  [account description amount date]
  {:account account :description description :amount amount :date date})

(defn- http-resp
  "Default OK HTTP response as String"
  [body]
  {:status 200
   :headers {"Content-Type" "application/json"}
   :body (if (map? body) (json/write-str body))})


(deftest invalid-address-test
  (is (= (app (mock/request :get "/invalid"))
         {:status 404, :body "Not found", :headers {"Content-Type" "text/html; charset=utf-8"}}))
  (is (= (app (mock/request :post "/invalid"))
         {:status 404, :body "Not found", :headers {"Content-Type" "text/html; charset=utf-8"}})))


;;;; NEW OPERATION TEST
(deftest one-new-operation-test
  (reset! ops {})
  (let [result {"1" [(build "Credit" "100.00" "15/10")]}
        response (app (mock/request :post "/new" (opp 1 "Credit" 100.0 "15/10")))]
    (is (= result @ops))
    (is (= response {:status 200, :headers {}, :body "ok"}))))


(deftest two-new-operations-test
  (reset! ops {})
  (let [result {"1" [(build "Debit"  -120.00 "16/10")
                     (build "Credit" 100.00 "15/10")]}
        response1 (app (mock/request :post "/new" (opp 1 "Credit" 100.0 "15/10")))
        response2 (app (mock/request :post "/new" (opp 1 "Debit" -120.0 "16/10")))]
    (is (= @ops result))
    (is (= response1 response2 {:status 200, :headers {}, :body "ok"}))))


(deftest new-operation-missing-parameters-test
  (let [m (opp 1 "Credit" 100.0 "15/10")]
    (testing "Missing account parameter"
      (let [result {:status 422, :body "Missing parameter: account", :headers {}}
            response (app (mock/request :post "/new" (dissoc m :account)))]
        (is (= result response))))

    (testing "Missing description parameter"
      (let [result {:status 422, :body "Missing parameter: description", :headers {}}
            response (app (mock/request :post "/new" (dissoc m :description)))]
        (is (= result response))))

    (testing "Missing date parameter"
      (let [result {:status 422, :body "Missing parameter: date", :headers {}}
            response (app (mock/request :post "/new" (dissoc m :date)))]
        (is (= result response))))

    (testing "Missing amount parameter"
      (let [result {:status 422, :body "Missing parameter: amount", :headers {}}
            response (app (mock/request :post "/new" (dissoc m :amount)))]
        (is (= result response))))

    (testing "Missing two parameters"
      (let [result1 {:status 422, :body "Missing parameter: amount", :headers {}}
            result2 {:status 422, :body "Missing parameter: date", :headers {}}
            response (app (mock/request :post "/new" (dissoc m :amount :date)))]
        (is (or (= result1 response) (= result2 response)))))))


;;;; CURRENT BALANCE TEST
(deftest current-balance-from-one-credit-operation-test
  (reset! ops {})
  (let [result (http-resp {:balance 100.00})
        _ (app (mock/request :post "/new" (opp 1 "Credit" 100.0 "15/10" )))
        response (app (mock/request :post "/balance" {:account 1}))]
    (is (= result response))))


(deftest current-balance-from-one-debit-operation-test
  (reset! ops {})
  (let [result (http-resp {:balance -120.00})
        _ (app (mock/request :post "/new" (opp 1 "Credit" -120.0 "15/10" )))
        response (app (mock/request :post "/balance" {:account 1}))]
    (is (= result response))))


(deftest current-balance-from-multiple-operations-test
  (reset! ops {})
  (let [result (http-resp {:balance 1201.00})
        ops [(opp 1 "Credit"  100.50  "15/10")   ;  100.50
             (opp 1 "Debit"  -120.00  "16/10" )  ; -19.50
             (opp 1 "Deposit" 220.50  "17/10" )  ;  201.00
             (opp 1 "Salary"  1000.00 "18/10" )]];  1201.00
    (doseq [op ops] (app (mock/request :post "/new" op)))
    (is (= result (app (mock/request :post "/balance" {:account 1}))))))


(deftest current-balance-multiple-accounts-test
  (reset! ops {})
  (let [result1 (http-resp {:balance 1100.50})
        result2 (http-resp {:balance -319.99})
        result3 (http-resp {:balance 5000.75})
        result4 {:status 200 :headers {} :body nil}
        ops [(opp 1 "Credit"    100.50    "15/10")    ; 1: 100.50
             (opp 2 "Debit"    -120.0     "15/10")    ; 2: -120.0
             (opp 2 "Purchase" -199.99    "20/10")    ; 2: -319.99
             (opp 3 "Salary"    5000.7565 "12/10")    ; 3: 5000.75
             (opp 1 "Salary"    1000.0    "18/10")]]  ; 1: 1100.50
    (doseq [op ops] (app (mock/request :post "/new" op)))

    (is (= result1 (app (mock/request :post "/balance" {:account 1}))))
    (is (= result2 (app (mock/request :post "/balance" {:account 2}))))
    (is (= result3 (app (mock/request :post "/balance" {:account 3}))))
    (is (= result4 (app (mock/request :post "/balance" {:account 4}))))))


(deftest current-balance-missing-parameters-test
  (reset! ops {})
  (app (mock/request :post "/new" (opp 1 "Salary" 1000.0 "18/10")))
  (let [result {:status 422 :headers {} :body "Missing parameter: account"}
        response (app (mock/request :post "/balance" {}))]
    (is (= result response))))


;;;; BANK STATEMENT TESTS
(deftest bank-statement-server-test
  (reset! ops {})
  (let [operations [(opp 2 "Salary"            5000.00 "05/08")
                    (opp 4 "Deposit"           1250.20 "02/08")
                    (opp 1 "Debit"            -999.99  "06/08")
                    (opp 2 "Purchase on eBay" -30.50   "01/09")
                    (opp 4 "Withdrawal"       -10.00   "01/09")
                    (opp 4 "Credit"            500.00  "25/08")
                    (opp 1 "Deposit from Ann"  1200.00 "16/09")
                    (opp 1 "Purchase a burger"-21.99   "16/09")
                    (opp 2 "Salary"            5000.00 "05/09")]]
    (doseq [op operations] (app (mock/request :post "/new" op)))

    (testing "Bank statement for account 1 from 06/08 to 16/09"
      (let [response (app (mock/request :post "/statement"
                                        {:account 1 :start "06/08" :end "16/09"}))
            result {"06/08" {"balance" -999.99
                             "operations" ["- Debit 999.99"]}
                    "16/09" {"balance" 178.02
                             "operations" ["- Purchase a burger 21.99"
                                           "- Deposit from Ann 1200.00"]}}
            body (json/read-str (:body response))]
        (is (= body result))))

    (testing "Bank statement for account 1 from 06/08 to 15/09"
      (let [response (app (mock/request :post "/statement"
                                        {:account 1 :start "06/08" :end "15/09"}))
            result {"06/08" {"balance" -999.99
                             "operations" ["- Debit 999.99"]}}
            body (json/read-str (:body response))]
        (is (= body result))))

    (testing "Bank statement for account 1 from 07/08 to 16/09"
      (let [response (app (mock/request :post "/statement"
                                        {:account 1 :start "07/08" :end "16/09"}))
            result {"16/09" {"balance" 178.02
                             "operations" ["- Purchase a burger 21.99"
                                           "- Deposit from Ann 1200.00"]}}
            body (json/read-str (:body response))]
        (is (= body result))))

    (testing "Bank statement for account 1 from 07/08 to 15/09"
      (let [response (app (mock/request :post "/statement"
                                        {:account 1 :start "07/08" :end "15/09"}))
            result {}
            body (json/read-str (:body response))]
        (is (= body result))))

    (testing "Bank statement for account 4 from 02/08 to 01/09"
      (let [response (app (mock/request :post "/statement"
                                        {:account 4 :start "02/08" :end "01/09"}))
            result {"02/08" {"balance" 1250.20
                             "operations" ["- Deposit 1250.20"]}
                    "25/08" {"balance" 1750.20
                             "operations" ["- Credit 500.00"]}
                    "01/09" {"balance" 1740.20
                             "operations" ["- Withdrawal 10.00"]}}
            body (json/read-str (:body response))]
        (is (= body result))))

    (testing "Bank statement for account 4 from 10/08 to 29/08"
      (let [response (app (mock/request :post "/statement"
                                        {:account 4 :start "10/08" :end "29/08"}))
            result {"25/08" {"balance" 1750.20
                             "operations" ["- Credit 500.00"]}}
            body (json/read-str (:body response))]
        (is (= body result))))

    (testing "Bank statement for account 4 from 20/08 to 10/09"
      (let [response (app (mock/request :post "/statement"
                                        {:account 4 :start "20/08" :end "10/09"}))
            result {"25/08" {"balance" 1750.20
                             "operations" ["- Credit 500.00"]}
                    "01/09" {"balance" 1740.20
                             "operations" ["- Withdrawal 10.00"]}}
            body (json/read-str (:body response))]
        (is (= body result))))))


(deftest bank-statement-missing-parameters-test
  (testing "Missing parameter account"
    (is (= (app (mock/request :post "/statement" {:start "10/10" :end "24/12"}))
           {:status 422 :headers {} :body "Missing parameter: account"})))

  (testing "Missing parameter start"
    (is (= (app (mock/request :post "/statement" {:account 1 :end "24/12"}))
           {:status 422 :headers {} :body "Missing parameter: start"})))

  (testing "Missing parameter end"
    (is (= (app (mock/request :post "/statement" {:start "10/10" :account 1}))
           {:status 422 :headers {} :body "Missing parameter: end"})))

  (testing "Missing parameter account"
    (is (= (app (mock/request :post "/statement" {}))
           {:status 422 :headers {} :body "Missing parameter: account"}))))


;;;; DEBT PERIODS TEST
(deftest debt-periods-server-test
  (reset! ops {})
  (let [ops [(opp 1 "Credit"    175.34   "01/01")
             (opp 1 "Debit"    -200.00   "03/01")
             (opp 1 "Salary"    50173.48 "10/01")
             (opp 2 "Salary"    6837.56  "10/01")
             (opp 2 "Purchase" -200.23   "14/01")
             (opp 2 "Purchase" -31.23    "14/01")
             (opp 2 "Debit"    -1123.00  "02/02")
             (opp 3 "Debit"    -1520.00  "05/01")
             (opp 3 "Purchase" -154.00   "19/01")
             (opp 3 "Debit"    -140.33   "22/02")]]
    (doseq [op ops] (app (mock/request :post "/new" op)))

    (testing "Account number 1"
      (let [response (app (mock/request :post "/debt" {:account 1}))
            result {"debts"
                    [{"start" "03/01" "end" "09/01" "principal" -24.65}]}
            body (json/read-str (:body response))]
        (is (= body result))))

    (testing "Account number 2"
      (let [response (app (mock/request :post "/debt" {:account 2}))
            result {"debts" []}
            body (json/read-str (:body response))]
        (is (= body result))))

    (testing "Account number 3"
      (let [response (app (mock/request :post "/debt" {:account 3}))
            result {"debts"
                    [{"start" "22/02" "principal" -1814.33}
                     {"start" "19/01" "end" "21/02" "principal" -1674.00}
                     {"start" "05/01" "end" "18/01" "principal" -1520.00}]}
            body (json/read-str (:body response))]
        (is (= body result))))))

(deftest debt-periods-missing-parameters-test
  (testing "Missing parameter account"
    (is (= (app (mock/request :post "/debt" {}))
           {:status 422 :headers {} :body "Missing parameter: account"}))))


src/balances/core.clj
(ns balances.core
  "Core functions for the bank operations.
  All functions are pure."
  (:require [balances.util :as util]))

;;;; Record definition
(defrecord Operation [description amount date]
  Object
  (toString [this]
    (str description " " (format "%.2f" (util/abs amount)))))

(defn build
  "Builds a Operation instance from description, amount and date"
  [description amount date]
  {:pre [(instance? String description)]}
  (Operation. description (util/to-format amount) (util/string->date date)))


;;;; Helper function
(defn- all-date->string
  "Convert the dates of each element from a seq of Operations from Date to
  String"
  [coll]
  (map #(update % :date util/date->string) coll))

(defn- compute-balances
  "Compute the balance for each date and sort the vector by date"
  [ops account]
  {:pos [vector?]}
  (letfn [(sum-last [val arr]
            (util/to-format (+ val (if (empty? arr) 0 (-> arr last :balance)))))
          (compress [a [k v]]
            (conj a {:date k :amount (apply + (map :amount v))}))
          (build-sum [a {:keys [date amount]}]
            (conj a {:date (util/date->string date) :balance (sum-last amount a)}))]

    (->> (group-by :date (ops account))
         (reduce compress [])
         (sort-by :date)
         (reduce build-sum []))))

(defn- compute-each-balance
  "Compute the balance for each date and return a map from date to balance"
  [ops account]
  {:pos [map?]}
  (let [compress (fn [m {:keys [balance date]}] (assoc m date balance))]
    (reduce compress {} (compute-balances ops account))))


;;;; Available functions
(defn new-operation
  "Returns a new map with a new Operation"
  [ops {:keys [account description amount date]}]
  {:pre [ops account description amount date]
   :pos [map?]}
  (update ops account #(conj % (build description amount date))))

(defn current-balance
  "Current balance of an account calculated from the sum of all previous
  operations"
  [ops account]
  {:pre [ops account]
   :pos [map?]}
  (if (contains? ops account)
    {:balance (->> (ops account) (map :amount) (apply +) util/to-format)}))

(defn bank-statement
  "Log of operations of an account between two dates"
  [ops account start-date end-date]
  {:pre [ops account start-date end-date]
   :pos [map?]}
  (let [within? (comp (util/within? start-date end-date) :date)
        each-balance (compute-each-balance ops account)]
    (reduce
      (fn [m [k v]] (assoc m k {:operations (map str (repeat "- ") v)
                                :balance (each-balance k)}))
      {} (->> (ops account) (filter within?) (all-date->string) (group-by :date)))))

(defn debt-periods
  "Map with a vector of periods when the account had a negative balance.
  Each element of the vector is a map with start date, negative balance and
  end date if the current balance is non negative"
  [ops account]
  {:pre [ops account]
   :pos [map?]}
  {:debts (let [mconj (fn [coll & xs] (apply conj coll (filter map? xs)))
                [head & tail] (drop-while (comp not neg? :balance)
                                          (compute-balances ops account))]
            (if head
              (reduce
                (fn [[head# & tail#] {:keys [balance date]}]
                  (if (= balance (:principal head#))
                    (conj tail# head#)
                    (mconj tail#
                           (if (contains? head# :end)
                             head#
                             (assoc head# :end (util/previous-day date)))
                           (if (neg? balance)
                             {:start date :principal balance}))))
                [{:start (:date head) :principal (:balance head)}]
                tail)
              []))})

src/balances/server.clj
(ns balances.server
  "Server functions.
  A state variable is made necessary to hold all past operations."
  (:require
    [compojure.route :as route]
    [ring.middleware.reload :as reload]
    [ring.util.response :as res]
    [ring.middleware.params :refer [wrap-params]]
    [ring.middleware.json-response :refer [wrap-json-response]]
    [ring.middleware.keyword-params :refer [wrap-keyword-params]]
    [ring.adapter.jetty :refer [run-jetty]]
    [compojure.core :refer [defroutes GET POST]]
    [balances.core :refer [new-operation current-balance
                           bank-statement debt-periods]]))

(def ops
  "State variable.
  ops is defined as an atom because it needs to be synchronous and there is
  no need to be coordinated"
  (atom {}))

(defn- validate
  "Validates the parameters of a function, since all values are necessary to
  properly continue"
  [m type]
  (let [fields {:balance   [:account]
                :new       [:account :amount :description :date],
                :debt      [:account]
                :statement [:account :start :end]}]
    (->> type fields (drop-while m) first)))

(defn- handler
  [func params]
  (if-let [miss (validate params func)]
    (res/status (res/response (str "Missing parameter: " (name miss))) 422)
    (let [{:keys [account start end]} params]
      (res/response
        (case func
          :new       (do (swap! ops new-operation params) "ok") ; side-effect: update ops
          :balance   (current-balance @ops account)
          :statement (bank-statement @ops account start end)
          :debt      (debt-periods @ops account))))))

(defroutes app-routes
           (POST "/new"       [& params] (handler :new       params))
           (POST "/balance"   [& params] (handler :balance   params))
           (POST "/statement" [& params] (handler :statement params))
           (POST "/debt"      [& params] (handler :debt      params))
           (route/not-found "Not found"))

(def app
  (-> app-routes
      (reload/wrap-reload)
      (wrap-keyword-params)
      (wrap-params)
      (wrap-json-response)))

(defn -main []
  (run-jetty app {:port 3000}))